docs/blog/2019-03-03-new-schema-customization/index.md

```diff
@@ -0,0 +1,270 @@
+
+On the other hand, `createResolvers` is to add _additional functionality_ to types. `createResolvers` also allows adding new _root fields_ to Query type.
+
+## `createTypes`
+
+Let's consider an example with `gatsby-source-filesystem`, where we are loading data from `authors.json` file. It has the following contents:
+
+```json
+[
+  {
+    "name": "Mikhail Novikov",
+    "birthday": "1987-09-25"
+  }
+]
+```
+
+This would be inferred in Gatsby as a Node type with a Date type for birthday.
+
+```graphql
+type AuthorsJson implements Node {
+  # default gatsby node fields
+  id: ID!
+  parent: Node
+  children: [Node!]!
+  internal: `Internal`,
+  # inferred fields
+  name: String
+  birthday: Date
+}
+```
+
+However this can break if we accidentally add invalid date as a birthday for a new node.
+
+```json
+[
+  {
+    "name": "Mikhail Novikov",
+    "birthday": "1987-09-25"
+  },
+  {
+    "name": "Stefan Probst",
+    "birthday": "Unknown"
+  }
+]
+```
+
+Now there is a type conflict between date and string and this will be inferred as string, possibly breaking our queries.
+
+````graphql
+```graphql
+type AuthorsJson implements Node {
+  # default gatsby node fields
+  id: ID!
+  parent: Node
+  children: [Node!]!
+  internal: `Internal`,
+  # inferred fields
+  name: String
+  birthday: String
+}
+````
+
+Luckily, now we can use `createTypes` action to force birthday to be a Date.
+
+```js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    # One must say that the type is a Node
+    type AuthorJson implements Node {
+      # However Node fields are optional and you don't have to add them
+      name: String
+      birthday: Date
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+Gatsby will now know that you want a Date and not override it with a string.
+
+You can specify types for some or all of the fields that you have on the given node type. Gatsby will add missing fields. This behaviour can be controlled with `@infer` and `@dontInfer` directives.
+
+```graphql
+# For this type `name` won't be added
+@dontInfer()
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+# For this type `name` won't be added but birtday won't get a Gatsby resolver for date formatting
+@dontInfer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+
+# For this type all fields `name`. Current default behaviour, but allows one to be explicit about it.
+@infer()
+type AuthorJson implements Node {
+  id: ID!
+}
+
+# Birthday will be Date, but we won't add a Gatsby resolver for date formatting
+@infer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  id: ID!
+}
+```
+
+## `createResolvers`
+
+This is a similar API to `setFieldsOnNodeType` in that it allows to add new fields and resolvers for them. However, this one is ran last, so you'd have all the schema available to you in it. Also it's possible to extend types like `Query` to add more root resolvers.
```

Comment:
Might be also worth mentioning that we *do* allow modifying types from third-party schemas.